src/java/picard/sam/markduplicates/MarkDuplicates.java

```diff
@@ -461,24 +463,33 @@ private void buildSortedReadEndLists(final boolean useBarcodes) {
                                 ((ReadEndsForMarkDuplicatesWithBarcodes) pairedEnds).readTwoBarcode = getReadTwoBarcodeValue(rec);
                         }
 
-                        // If the second read is actually later, just add the second read data, else flip the reads
-                        if (sequence > pairedEnds.read1ReferenceIndex ||
-                                (sequence == pairedEnds.read1ReferenceIndex && coordinate >= pairedEnds.read1Coordinate)) {
-                            pairedEnds.read2ReferenceIndex = sequence;
-                            pairedEnds.read2Coordinate = coordinate;
+                        // If the other read is actually later, simply add the other read's data as read2, else flip the reads
+                        if (matesRefIndex > pairedEnds.read1ReferenceIndex ||
+                                (matesRefIndex == pairedEnds.read1ReferenceIndex && matesCoordinate >= pairedEnds.read1Coordinate)) {
+                            pairedEnds.read2ReferenceIndex = matesRefIndex;
+                            pairedEnds.read2Coordinate = matesCoordinate;
                             pairedEnds.read2IndexInFile = indexForRead;
                             pairedEnds.orientation = ReadEnds.getOrientationByte(pairedEnds.orientation == ReadEnds.R,
                                     rec.getReadNegativeStrandFlag());
                         } else {
                             pairedEnds.read2ReferenceIndex = pairedEnds.read1ReferenceIndex;
                             pairedEnds.read2Coordinate = pairedEnds.read1Coordinate;
                             pairedEnds.read2IndexInFile = pairedEnds.read1IndexInFile;
-                            pairedEnds.read1ReferenceIndex = sequence;
-                            pairedEnds.read1Coordinate = coordinate;
+                            pairedEnds.read1ReferenceIndex = matesRefIndex;
+                            pairedEnds.read1Coordinate = matesCoordinate;
                             pairedEnds.read1IndexInFile = indexForRead;
                             pairedEnds.orientation = ReadEnds.getOrientationByte(rec.getReadNegativeStrandFlag(),
                                     pairedEnds.orientation == ReadEnds.R);
                         }
+                        // if the two read ends are in the same position, pointing in opposite directions,
+                        // the orientation is undefined and the procedure above
+                        // will depend on the order of the reads in the file.
+                        // To avoid this, we set it explicitly (to FR):
+                        if (pairedEnds.read2ReferenceIndex == pairedEnds.read1ReferenceIndex &&
+                                pairedEnds.read2Coordinate == pairedEnds.read1Coordinate &&
+                                pairedEnds.orientation == ReadEnds.RF) {
+                            pairedEnds.orientation = ReadEnds.FR;
```

Comment:
I think this is confusing to be separate from the `if/else` starting on line 467.  Can you make it part of the first `if` on line 467, or make it its own `if (sequence == pairedEnds.read1ReferenceIndex && coordinate == pairedEnds.read1Coordinate)` and then you can have `if/else if/else`?  I would prefer the former.